src/Compiler/Optimize/Optimizer.fs

```diff
@@ -3054,17 +3063,25 @@ and TryOptimizeVal cenv env (vOpt: ValRef option, mustInline, inlineIfLambda, va
     | ConstExprValue(_size, expr) ->
         Some (remarkExpr m (copyExpr g CloneAllAndMarkExprValsAsCompilerGenerated expr))
 
-    | CurriedLambdaValue (_, _, _, expr, _) when mustInline || inlineIfLambda ->
-        let exprCopy = CopyExprForInlining cenv inlineIfLambda expr m
-        Some exprCopy
+    | CurriedLambdaValue (_, _, _, expr, _) when shouldInline || inlineIfLambda ->
+        let fvs = freeInExpr CollectLocals expr
+        if fvs.UsesMethodLocalConstructs then
+            // Discarding lambda for binding because uses protected members --- TBD: Should we warn or error here
+            None 
+        elif fvs.FreeLocals.ToArray() |> Seq.fold(fun acc v -> if not acc then v.Accessibility.IsPrivate else acc) false then
```

Comment:
Isn't this:


```suggestion
        elif fvs.FreeLocals |> Seq.exists(fun v -> v.Accessibility.IsPrivate ) then
```

Reply:
True ...

but the actual code in main is:

````
    | CurriedLambdaValue (_, _, _, expr, _) when mustInline || inlineIfLambda ->
        let exprCopy = CopyExprForInlining cenv inlineIfLambda expr m
        Some exprCopy
````
This replacement code just expands that to eliminate two causes of inlining problems.
MethodLocalConstructs and Local value useage that are private

````
    | CurriedLambdaValue (_, _, _, expr, _) when shouldInline || inlineIfLambda ->
        let fvs = freeInExpr CollectLocals expr
        if fvs.UsesMethodLocalConstructs then
            // If the lambda expression uses MethodLocal constructs then don't inline
            None 
        elif fvs.FreeLocals.ToArray() |> Seq.fold(fun acc v -> if not acc then v.Accessibility.IsPrivate else acc) false then
            // If the lambdo expression uses uses a private binding then don't inline
            None
        else
            let exprCopy = CopyExprForInlining cenv inlineIfLambda expr m
            Some exprCopy
````

Reply:
I get that.

This was more of a stylistic feedback - the current `Seq.fod` with a bool state feels like it could be simplified down to a `fvs.FreeLocals.exists` call.